### PR TITLE
issue 8185: fix for allowing spec name with square brackets

### PR DESCRIPTION
**Pull request by vmwhereqa**
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

File: packages/wdio-config/src/lib/ConfigParser.ts [hunk 1]
```diff
@@ -248,7 +248,7 @@ export default class ConfigParser {
             }
         }
 
-        return this.filterSpecs(specs, <string[]>exclude)
+        return this.filterSpecs([...spec, ...specs], <string[]>exclude)
```
Comment (christian-bromann) [inline packages/wdio-config/src/lib/ConfigParser.ts:251]: I think this change is unrelated to the issue, we can revert it which also should make the unit test pass.

File: packages/wdio-config/tests/FileSystemPathService.test.ts [hunk 2]
```diff
@@ -14,7 +14,6 @@ vi.mock('glob', () => ({
 
 describe('FileSystemPathService', () => {
     afterEach(() => {
-        (glob.sync as MockedFunction<any>).mockClear()
```
Comment (christian-bromann) [inline packages/wdio-config/tests/FileSystemPathService.test.ts:17]: I think it makes sense to keep this and better rewrite this as follows:

```ts
vi.mocked(glob.sync).mockClear()
```

**Review by christian-bromann — COMMENTED** [review]
Great work, some comments!

**Review by christian-bromann — APPROVED** [review]
LGTM 👍

**Comment by christian-bromann** [comment]
@vmwhereqa it seems the unit test added fails 🤔

**Comment by vmwhereqa** [comment]
> @vmwhereqa it seems the unit test added fails 🤔

let me check

File: packages/wdio-config/src/lib/FileSystemPathService.ts [hunk 3]
```diff
@@ -31,7 +31,20 @@ export default class FileSystemPathService implements PathService {
...
+        const fileName = pattern.startsWith('/') ? pattern : path.resolve(this.getcwd(), pattern)
+        /**
+         * given that glob treats characters like `[` or `{` in a special way
+         * and we also want to be able to find files with these characters included
+         * we add an additional check to see if the file as pattern exists.
+         * add file to globResult only if filename doesn't include pattern(*)
+         * and globResult doest contain the fileName
+         * and file should be available
+         */
+        if (!pattern.includes('*') && !globResult.includes(fileName) && fs.existsSync(fileName)) {
```
Comment (vmwhereqa) [inline packages/wdio-config/src/lib/FileSystemPathService.ts:44]: @christian-bromann : In my test I am trying to make `fs.existsSync(fileName)` always retrun _true_  using vi.mock but this change is getting applied to all other test. Is there way to make this module mocking specific to test block I have written.

**Comment by vmwhereqa** [comment]
@christian-bromann: Newly added unit test case is passing but the test `should return true if file esxists` is failing due to mocking only `existsSync`. Tried multiple ways but didn't workout. Could you please help me in fixing the failing test by not applying `fs` mocking to the failing test

File: packages/wdio-config/tests/FileSystemPathService.test.ts [hunk 4]
```diff
@@ -28,66 +33,72 @@ describe('FileSystemPathService', () => {
...
-            var svc = new FileSystemPathService()
+            const svc = new FileSystemPathService()
             expect(() => svc.getcwd()).toThrowError('Unable to find current working directory from process')
         })
     })
 
     describe('isFile', function () {
         it('should return true if file esxists', function () {
-            var svc = new FileSystemPathService()
+            const svc = new FileSystemPathService()
```
Comment (christian-bromann) [inline packages/wdio-config/tests/FileSystemPathService.test.ts:49]:
```suggestion
            const svc = new FileSystemPathService()
            vi.mocked(fs.existsSync).mockReturnValue(true)
```

Comment (vmwhereqa) [reply packages/wdio-config/tests/FileSystemPathService.test.ts:49]: yes that works, is there a way to ignore `fs.lstatSync(filepath)` from mocking

File: packages/wdio-config/tests/FileSystemPathService.test.ts [hunk 5]
```diff
@@ -28,66 +33,72 @@ describe('FileSystemPathService', () => {
...
             expect(svc.isFile(INDEX_PATH)).toBeTruthy()
         })
 
         it("should return false if file doesn't exist", function () {
-            var svc = new FileSystemPathService()
+            const svc = new FileSystemPathService()
```
[the remark this answers is not in the document]
Comment (vmwhereqa) [reply packages/wdio-config/tests/FileSystemPathService.test.ts:101]: > We need to define return values for places where `fs` is being used elsewhere in the file. Try working with this.

makes sense. I have implemented for all fs functions in the file

File: packages/wdio-config/tests/FileSystemPathService.test.ts [hunk 6]
```diff
@@ -28,66 +37,115 @@ describe('FileSystemPathService', () => {
...
+                },
+                isSymbolicLink: function (): boolean {
+                    throw new Error('Function not implemented.')
+                },
+                isFIFO: function (): boolean {
+                    throw new Error('Function not implemented.')
+                },
+                isSocket: function (): boolean {
+                    throw new Error('Function not implemented.')
+                }
```
Comment (vmwhereqa) [inline packages/wdio-config/tests/FileSystemPathService.test.ts:93]: @christian-bromann : is there a better way to do this?

File: packages/wdio-config/tests/FileSystemPathService.test.ts [hunk 7]
```diff
@@ -28,66 +37,115 @@ describe('FileSystemPathService', () => {
...
+                },
+                isSymbolicLink: function (): boolean {
+                    return false
+                },
+                isFIFO: function (): boolean {
+                    return false
+                },
+                isSocket: function (): boolean {
+                    return false
+                }
```
Comment (vmwhereqa) [inline packages/wdio-config/tests/FileSystemPathService.test.ts:93]: @christian-bromann: is there a way to do this?

Comment (christian-bromann) [reply packages/wdio-config/tests/FileSystemPathService.test.ts:93]: there is no need to create a mock for the complete object, just check what this specific function is using, e.g. in this case:

```js
return (fs.existsSync(filepath) && fs.lstatSync(filepath).isFile())
```
so you only need `isFile`

**Review by christian-bromann — APPROVED** [review]
Great job 👍 
Thanks!
